src/braket/circuits/moments.py

```diff
@@ -257,13 +257,17 @@ def sort_moments(self) -> None:
         key_readout_noise = []
         moment_copy = OrderedDict()
         sorted_moment = OrderedDict()
+        last_measure = self._depth
 
         for key, instruction in self._moments.items():
             moment_copy[key] = instruction
             if key.moment_type == MomentType.READOUT_NOISE:
                 key_readout_noise.append(key)
             elif key.moment_type == MomentType.INITIALIZATION_NOISE:
                 key_initialization_noise.append(key)
+            elif key.moment_type == MomentType.MEASURE:
+                last_measure = key.time
+                key_noise.append(key)
```

Comment:
interesting, I wonder if this could have unindented consequences if we introduce mid-circuit measurement, because of `last_measure = key.time` and `max(last_measure)`.

Reply:
Good point, this likely will need to change for mid-circuit measurement because this is assuming there is no measure _before_ the readout noise is added. 

Maybe there is a way to make sure the `last_measure` is actually the last measure on the circuit?

Reply:
Similar to the mid-circuit measurement consequences, should the following circuit be allowed:

```
circ = Circuit().h(0).cnot(0, 1)
circ.apply_readout_noise(Noise.BitFlip(probability=0.1), target_qubits=1)
circ.measure([0])
circ.cnot(1, 2)
circ.measure(3)
```
```
T  : │  0  │  1  │        2        │  3  │
      ┌───┐       ┌───┐                   
q0 : ─┤ H ├───●───┤ M ├───────────────────
      └───┘   │   └───┘                   
            ┌─┴─┐       ┌─────────┐       
q1 : ───────┤ X ├───●───┤ BF(0.1) ├───────
            └───┘   │   └─────────┘       
                  ┌─┴─┐                   
q2 : ─────────────┤ X ├───────────────────
                  └───┘                   
                                    ┌───┐ 
q3 : ───────────────────────────────┤ M ├─
                                    └───┘ 
T  : │  0  │  1  │        2        │  3  │
```

Reply:
this is interesting, and maybe beyond the scope of this PR. 

The circuit you've shown applies readout noise to qubit 1, even through the measurement is on qubit 3. That's strange and should not work. The apply readout (i.e. measurement noise), should only apply _before_ a measurement operation. The reason it currently applies at the end is because we _implicitly_ added measurements to the end of a circuit. 

So 
circ.apply_readout_noise(..., target_qubits=1)
circ.measure([0])
should have no noise
But
circ.apply_readout_noise(..., target_qubits=0)
circ.measure([0])
should have the noise before the measure operation.

Reply:
You are right, I think this problem is probably worth it's own PR.

Maybe there should be some validation to check that a measure instruction exists on the readout target and we can keep track of all the measure locations to add the noise right before it.